src/hdnode.ts

```diff
@@ -0,0 +1,40 @@
+import {Buffer} from "buffer";
+import {assertValidPath} from "./utils";
+import BN from "bn.js";
+import {deriveChildSK} from "./key-derivation";
+
+export class HDNode {
+
+    public key: Buffer;
+    private depth: number;
+    private path: string;
+
+    private constructor(key: Buffer, depth: number, path: string) {
+       this.key = key;
+       this.depth = depth;
+       this.path = path;
+    }
+
+    private derive(): HDNode {
+        if(this.depth === 4) {
+            return this;
+        }
+        const index = this.getIndex(this.path, this.depth + 1);
+        return new HDNode(deriveChildSK(this.key, index), this.depth + 1, this.path).derive();
```

Comment:
Something about this seems strange. Can we do this functionally instead of using OOP?

eg:
```typescript
function derive(key: Buffer, path: string): Buffer;

function pathToIndices(path: string): BN[];
```

Reply:
not sure what your functions are doing.
Basically HDNode represents node in account tree and derive returns one node deeper (using path and depth to determine what node to derive).

Reply:
I don't think it makes sense to create objects in the middle of the path, but rather have an interface that just traverses the path and gives back a key at the end.
```typescript
function derive(key: Buffer, path: string): Buffer {
  let k = key;
  pathToIndices(path).forEach((i) => k = deriveChildSK(k, i));
  return k;
}
``

Reply:
Well yeah, that works for eth2 account hierarchy because it's pretty basic but other blockchains might decide to have
different processing on some depths.
Hint, look at eth1:
https://github.com/ethers-io/ethers.js/blob/master/src.ts/utils/hdnode.ts

Reply:
not sure whats best way to proceed, others might decided to use similar paths.

Reply:
I think going with this method is actually fine the section in etheres is [here](https://github.com/ethers-io/ethers.js/blob/master/src.ts/utils/hdnode.ts#L185).

Reply:
If you look at the [EIP]( https://github.com/ethereum/EIPs/blob/c658eb6cbc0dc7dd7edaa8db429cc9abcf450fa7/EIPS/eip-2334.md
) purpose/motivation sections, it seems like the intent is that different blockchains will use this standard as-is (ie: use the same depth).

Reply:
> ```ts
> function derive(key: Buffer, path: string): Buffer;
> 
> function pathToIndices(path: string): BN[];
> ```

Both of these functions make a lot of sense to me. We keep track of the path as an array of u32s. If there is a `'` (or `h`) that follows the integer in the string representation, the first bit of the binary encoding of the u32 gets flipped to a 1.
For `pathToIndices`, you do not need to return BNs since the spec defines u32s which can be represented in JS without loss of precision. See https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki

Using `h` over `'` is useful for bash

See https://github.com/bcoin-org/bcoin/blob/85ed59c842f2aa4c1a8154d74a9cad7696cdfee3/lib/hd/common.js#L52

Also:
https://github.com/bcoin-org/bcoin/blob/85ed59c842f2aa4c1a8154d74a9cad7696cdfee3/lib/hd/private.js#L171